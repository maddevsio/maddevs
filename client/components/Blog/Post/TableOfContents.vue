<template>
  <div class="table-of-contents">
    <div class="table-of-contents__title">
      Contents:
    </div>
    <Simplebar
      v-if="anchors && anchors.length"
      ref="anchors"
      class="table-of-contents__links"
    >
      <NuxtLink
        v-for="(anchor, i) in anchors"
        :key="anchor.lable"
        :to="anchor.link"
        class="table-of-contents__links-link"
        :class="{ 'table-of-contents__links-link--active': getActiveAnchor(anchor.link, i) }"
      >
        <span>{{ i + 1 }}.</span> {{ anchor.lable }}
      </NuxtLink>
    </Simplebar>
  </div>
</template>

<script>
import Simplebar from 'simplebar-vue'
import mainMixins from '@/mixins/mainMixins'

export default {
  name: 'TableOfContents',

  components: { Simplebar },

  mixins: [mainMixins],

  props: {
    slice: {
      type: Object,
      required: true,
    },
  },

  data() {
    return {
      activeAnchor: null,
      scrollBar: null,
    }
  },

  computed: {
    anchors() {
      if (this.slice && this.slice.items && !this.slice.items.length) return []
      return this.slice.items
        .filter(item => item.lable.length)
        .map(item => ({
          lable: item.lable[0].text,
          link: `#${this.createAnchorID(item.lable[0].text)}`,
        }))
    },
  },

  mounted() {
    const sections = document.querySelectorAll('.anchor_title')
    const config = {
      rootMargin: '-100px 0px -55%',
    }

    const observer = new IntersectionObserver(entries => {
      entries.forEach(entry => {
        if (entry.isIntersecting) {
          this.intersectionHandler(entry)
        }
      })
    }, config)

    sections.forEach(section => {
      observer.observe(section)
    })

    if (this.$refs && this.$refs.anchors && this.$refs.anchors.SimpleBar) {
      this.scrollBar = this.$refs.anchors.SimpleBar.getScrollElement()
    }
  },

  methods: {
    intersectionHandler(entry) {
      if (this.anchors.some(a => a.link.includes(entry.target.id))) {
        this.activeAnchor = `#${entry.target.id}`
      }
    },

    getActiveAnchor(link, index) {
      if (link === this.activeAnchor) {
        if (index === 4) {
          this.scrollBar.scrollTo({ top: 0, behavior: 'smooth' })
          return true
        }
        if (index === 5) {
          this.scrollBar.scrollTo({ top: 400, behavior: 'smooth' })
          return true
        }
        return true
      }
      return false
    },
  },
}
</script>

<style lang="scss" scoped>
.table-of-contents {
  width: 200px;
  max-width: 200px;

  &__title {
    color: #101113;
    line-height: 130%;
    margin-bottom: 18px;
    font-weight: bold;
  }

  &__links {
    width: 100%;
    max-height: 400px;
    display: flex;
    flex-direction: column;
    padding-right: 12px;

    &-link {
      width: 100%;
      display: flex;
      text-decoration: none;
      font-size: 14px;
      line-height: 130%;
      letter-spacing: -0.1px;
      color: #101113;
      margin-bottom: 6px;
      padding: 10px 8px;
      box-sizing: border-box;
      border-radius: 8px;

      span {
        margin-right: 6px;
      }

      &:hover {
        background-color: rgba(247, 199, 68, 0.1);
      }

      &--active {
        background-color: rgba(204, 160, 40, 0.2);
      }
    }
  }
}
</style>
