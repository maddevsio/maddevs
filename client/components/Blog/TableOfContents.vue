<template>
  <div class="table-of-contents">
    <div class="table-of-contents__title">Table of contents</div>
    <ol class="table-of-contents__list">
      <li v-for="(link, index) in content" :key="index" class="table-of-contents__list-item">
        <template v-if="link.type === 'list-item'">
          <a :href="link.spans.length ? link.spans[0].data.url : null" @click.prevent="scrollToSection">
            {{ link.text }}
          </a>
        </template>
      </li>
    </ol>
  </div>
</template>

<script>
export default {
  name: 'TableOfContents',
  props: {
    content: {
      type: Array,
      required: true,
    },
  },

  created() {
    this.content.forEach(link => {
      if (link.spans.length) {
        // eslint-disable-next-line
        link.spans[0].data.url = link.spans[0].data.url.replace('https://', '');
      }
    })
  },

  methods: {
    scrollToSection(e) {
      const id = e.target.getAttribute('href') ? e.target.getAttribute('href').replace('#', '') : null
      const element = document.getElementById(id)
      if (element !== null) {
        window.scrollTo({
          top: element.offsetTop + 180,
          behavior: 'smooth',
        })
      }
    },
  },
}
</script>

<style lang="scss" scoped>
@import '../../assets/styles/vars';
@import '../../assets/styles/cases/mixins';

.table-of-contents {
  padding: 23px 24px;
  background-color: $bgcolor--silver;

  &__title {
    @include title($text-color--black-cases, 21px, -0.04em);
    margin-bottom: 20px;
    font-family: 'Inter', sans-serif;
    font-weight: 400;
  }

  &__list {
    padding-left: 20px;
    margin: 0;

    &-item {
      padding: 7px 0 12px;
      font-family: 'Inter', sans-serif;
      font-weight: 400;

      a {
        text-decoration: none;
        color: $text-color--black-cases;

        &:hover {
          text-decoration: underline;
        }

        &:visited {
          color: inherit;
        }
      }
    }
  }
}
</style>
