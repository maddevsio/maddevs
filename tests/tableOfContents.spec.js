import { mount } from '@vue/test-utils'
import TableOfContents from '@/components/Blog/TableOfContents'

describe('table of contents block in single post page', () => {
  let wrapper

  const content = [
    {
      spans: [
        {
          data: {
            lunk_type: 'Web',
            url: 'https://#gallery',
          },
          start: 0,
          end: 7,
          type: 'hyperlink',
        },
      ],
      text: 'Gallery',
      type: 'list-item',
    },
    {
      spans: [
        {
          data: {
            lunk_type: 'Web',
            url: 'https://#main',
          },
          start: 0,
          end: 7,
          type: 'hyperlink',
        },
      ],
      text: 'Main',
      type: 'list-item',
    },
  ]

  beforeEach(() => {
    wrapper = mount(TableOfContents, {
      propsData: { content },
    })
  })

  // ------ IMPORTANT ----- //
  it('is a Vue instance', () => {
    expect(wrapper.exists()).toBeTruthy()
  })

  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
  // --------------------- //

  it('ordered list rendered', () => {
    expect(wrapper.find('ol').exists()).toBeTruthy()
  })

  it('link replaced https', () => {
    expect(wrapper.find('li').find('a').attributes('href')).toMatch(
      content[0].spans[0].data.url.replace('https://', ''),
    )
  })
})
