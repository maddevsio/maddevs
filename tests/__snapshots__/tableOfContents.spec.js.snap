// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table of contents block in single post page renders correctly 1`] = `
<div
  class="table-of-contents"
>
  <div
    class="table-of-contents__title"
  >
    Table of contents
  </div>
   
  <ol
    class="table-of-contents__list"
  >
    <li
      class="table-of-contents__list-item"
    >
      <a
        href="#gallery"
      >
        
          Gallery
        
      </a>
    </li>
    <li
      class="table-of-contents__list-item"
    >
      <a
        href="#main"
      >
        
          Main
        
      </a>
    </li>
  </ol>
</div>
`;
